### PR TITLE
[Bookmark] Conversion of wrapped anchors.

### PR DESCRIPTION
**Pull request by pszczesniak**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Internal: Conversion of wrapped anchors that contains `id` and/or `name` attribute.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-bookmark/src/bookmarkediting.ts [hunk 1]
```diff
@@ -92,6 +96,8 @@ export default class BookmarkEditing extends Plugin {
 		const { editor } = this;
 		const { conversion, t } = editor;
 
+		editor.data.upcastDispatcher.on<UpcastElementEvent>( 'element:a', dataViewModelAnchorInsertion( editor ) );
```
Comment (niegowski) [inline packages/ckeditor5-bookmark/src/bookmarkediting.ts:99]: I'd suggest using the official public API for conversion:
```suggestion
		conversion.for( 'upcast' ).add( dispatcher => dispatcher.on<UpcastElementEvent>( 'element:a', dataViewModelAnchorInsertion( editor ) );
```

File: packages/ckeditor5-bookmark/src/bookmarkediting.ts [hunk 2]
```diff
@@ -181,27 +178,90 @@ export default class BookmarkEditing extends Plugin {
...
+		}
+
+		conversionApi.consumable.consume( viewItem, match );
+
+		if ( anchorId === anchorName ) {
+			conversionApi.consumable.consume( viewItem, { attributes: [ 'name' ] } );
+		}
+
+		conversionApi.updateConversionResult( bookmark, data );
+		const { modelCursor, modelRange } = conversionApi.convertChildren( viewItem, data.modelCursor );
```
Comment (niegowski) [inline packages/ckeditor5-bookmark/src/bookmarkediting.ts:255]:
```suggestion
		conversionApi.updateConversionResult( bookmark, data );

		const { modelCursor, modelRange } = conversionApi.convertChildren( viewItem, data.modelCursor );
```

File: packages/ckeditor5-bookmark/src/bookmarkediting.ts (hunk 1 above)
Comment (pszczesniak) [reply packages/ckeditor5-bookmark/src/bookmarkediting.ts]: I've moved it also into right place.

File: packages/ckeditor5-bookmark/src/bookmarkediting.ts [hunk 3]
```diff
@@ -181,27 +180,94 @@ export default class BookmarkEditing extends Plugin {
...
+		data.modelCursor = modelCursor;
+		data.modelRange = modelWriter.createRange( data.modelRange!.start, modelRange!.end );
+	};
+}
+
+/**
+ * Normalize the bookmark configuration option `enableNonEmptyBookmarkConversion`.
+ */
+function isEnabledNonEmptyBookmarkConversion( editor: Editor ): boolean {
+	const enableNonEmptyBookmarkConversion = editor.config.get( 'bookmark.enableNonEmptyBookmarkConversion' );
```
Comment (Witoso) [inline packages/ckeditor5-bookmark/src/bookmarkediting.ts:269]: Name change after the refinement?

`enableNonEmptyAnchorConversion`

Comment (pszczesniak) [reply packages/ckeditor5-bookmark/src/bookmarkediting.ts:269]: Will be done in a follow-up.

File: packages/ckeditor5-bookmark/src/bookmarkediting.ts [hunk 4]
```diff
@@ -181,27 +180,94 @@ export default class BookmarkEditing extends Plugin {
...
+
+		conversionApi.consumable.consume( viewItem, match );
+
+		if ( anchorId === anchorName ) {
+			conversionApi.consumable.consume( viewItem, { attributes: [ 'name' ] } );
+		}
+
+		conversionApi.updateConversionResult( bookmark, data );
+
+		const { modelCursor, modelRange } = conversionApi.convertChildren( viewItem, data.modelCursor );
```
Comment (niegowski) [inline packages/ckeditor5-bookmark/src/bookmarkediting.ts:258]: Let's add a code comment here. This part uses the result of bookmark insertion to convert the anchor content after the bookmark element (not inside it).

**Review by illia-stv — APPROVED** [review]
✅

**Review by mmotyczynska — APPROVED** [review]

File: packages/ckeditor5-bookmark/tests/manual/bookmark.html [hunk 5]
```diff
@@ -17,12 +19,226 @@
...
+	<p>
+		<a id="foo" name="foo"></a>pointed with the same id and name
+	</p>
+	<p>
+		<a id="foo"></a>pointed only with the id
+	</p>
+</div>
+
+
+
```
[the remark this answers is not in the document]
Comment (pszczesniak) [reply packages/ckeditor5-bookmark/tests/manual/bookmark.html:241]: Good catch 👍  will remove those empty lines.

File: packages/ckeditor5-bookmark/tests/bookmarkediting.js [hunk 6]
```diff
@@ -387,127 +389,345 @@ describe( 'BookmarkEditing', () => {
...
+					editor.setData( '<p><a id="foo" name="bar">foobar</a></p>' );
+
+					expect( getModelData( model, { withoutSelection: true } ) ).to.equal(
+						'<paragraph>foobar</paragraph>'
+					);
+				} );
+			} );
 		} );
 
 		describe( 'with GHS enabled', () => {
```
[the remark this answers is not in the document]
Comment (pszczesniak) [reply packages/ckeditor5-bookmark/tests/bookmarkediting.js:737]: Now it's tested correctly :point\_right: 63c195f.

**Review by niegowski — APPROVED** [review]

File: packages/ckeditor5-bookmark/src/bookmarkediting.ts (hunk 3 above)
Comment (pszczesniak) [reply packages/ckeditor5-bookmark/src/bookmarkediting.ts:271]: :point\_right: #17228

**Pull request merged**